File: 3.0/golang/34.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func dfs(graph [][]int, colors []int, now int) {
	colors[now] = 1

	for _, neig := range graph[now] {
		if colors[neig] == 0 {
			dfs(graph, colors, neig)
		} else if colors[neig] == 1 {
			flag = false
			break
		}
	}

	colors[now] = 2
	ans = append(ans, now)
}

var flag = true
var ans []int

func main() {
	var N, M int
	in := bufio.NewReader(os.Stdin)
	fmt.Fscanf(in, "%d %d\n", &N, &M)

	graph := make([][]int, N+1)
	colors := make([]int, N+1)

	var v, e int
	for i := 0; i < M; i++ {
		fmt.Fscanf(in, "%d %d\n", &v, &e)
		graph[v] = append(graph[v], e)
	}

	fmt.Println(graph)
	for i := 1; i < N+1; i++ {
		if colors[i] == 0 {
			dfs(graph, colors, i)
		}
	}

	if flag {
		for i := len(ans) - 1; i >= 0; i-- {
			fmt.Print(ans[i], " ")
		}
	} else {
		fmt.Println(-1)
	}
}

File: 3.0/python/34.py
from sys import setrecursionlimit
setrecursionlimit(1000000)

n, m = map(int, input().split())
graph = [[] for _ in range(n+1)]
colors = [0 for _ in range(n+1)]
ans = []

for _ in range(m):
    v, e = map(int, input().split())
    graph[v].append(e)

def dfs(graph, colors, now):
    global flag
    colors[now] = 1

    for neig in graph[now]:
        if colors[neig] == 0:
            dfs(graph, colors, neig)
        elif colors[neig] == 1:
            flag = False
            break

    colors[now] = 2
    ans.append(now)

flag = True
for i in range(1, n+1):
    if colors[i] == 0:
        dfs(graph, colors, i)

if flag:
    print(*reversed(ans))
else:
    print(-1)